### PR TITLE
Delete Notifications When a post is deleted

### PR DESCRIPTION
**Pull request by terabin**

### CODE REVIEW

**Comment by tobscure** [comment]
Thanks for the amended PR. Still not quite right as notifications can have subjects that are not posts, so deleting all by ID like this won't work. But getting there. The solution is slightly more complex, I will have a think about the best way to do it.

**Comment by terabin** [comment]
So, maybe the best way to do this is in each extension?

**Comment by tobscure** [comment]
Nope the best way is still in core, but will probably involve looping through each registered notification type and detecting whether or not the subject is a post. I'll work on it when I have some time.

**Comment by terabin** [comment]
It's just PostLiked, PostMentioned and userMentioned?

**Comment by terabin** [comment]
But each notification type have a different text, so, it's not possible get info about post type in each extension, right? Just curious.

**Comment by terabin** [comment]
I improved the post code, and worked for now, can you confirm? Just thinking about discussion.

**Comment by renyuneyun** [comment]
I guess coding style (StyleCI check) is the reason why this isn't merged (?)

**Comment by franzliedke** [comment]
No, Toby will need to review the logic. I can make some comments on the PHP code really quick...

File: src/Core/Post.php [hunk 1]
```diff
@@ -87,6 +88,21 @@ public static function boot()
...
+            //Delete notifications about this post
+            $notifs = Notification::where('subject_id', $post->id)->orWhere('subject_id', $post->discussion_id)->get();
+
+            foreach ($notifs as $notif)
+            {
+                if (!$notif->subject) {
+                    $notif->delete();
+                }
+                else {
+                    if ($notif->type == $post->type) {
```
Comment (franzliedke) [inline src/Core/Post.php]: You can "collapse" this into one `else if` statement.

File: src/Core/Post.php [hunk 2]
```diff
@@ -87,6 +88,21 @@ public static function boot()
...
+
+            //Delete notifications about this post
+            $notifs = Notification::where('subject_id', $post->id)->orWhere('subject_id', $post->discussion_id)->get();
+
+            foreach ($notifs as $notif)
+            {
+                if (!$notif->subject) {
+                    $notif->delete();
+                }
+                else {
```
Comment (franzliedke) [inline src/Core/Post.php]: ...and this can be on one line: `} else if (...) {`.

Comment (Luceos) [inline src/Core/Post.php]: @franzliedke meant `} else {`

**Comment by tobscure** [comment]
The notification subject type checking should be done using the Blueprint for the notification type... it might be a little complex and requires some knowledge of how the (undocumented) notification system works, so might be easier to just leave for now... do we have an issue for this?

**Comment by franzliedke** [comment]
We have issue #1025, I will reference your comment there. :)

**Comment by tobscure** [comment]
In any case, thanks for your efforts on this one @terabin. Sorry we didn't adopt your solution — as I said above, this problem is more complex than it seems — but we do appreciate the PR!

**Pull request closed without merging**